### ISSUE TITLE
Upgrade scripts and tests should take arbitrary published versions

### ISSUE THREAD

**Issue by ihmccreery** [kubernetes/kubernetes#15506]
Right now the [cluster_upgrade e2e](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/cluster_upgrade.go) takes:

```
// realVersion turns a version constant s--one accepted by cluster/gce/upgrade.sh--
// into a deployable version string. If the s is not known to be a version
// constant, it will assume it is already a valid version, and return s directly.
//
// NOTE: KEEP THIS LIST UP-TO-DATE WITH THE CODE BELOW.
// The version strings supported are:
// - "latest_stable"   (returns a string like "0.18.2")
// - "latest_release"  (returns a string like "0.19.1")
// - "latest_ci"       (returns a string like "0.19.1-669-gabac8c8")
```

And the [upgrade.sh](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/upgrade.sh) script takes similar params (via [set_binary_version](https://github.com/kubernetes/kubernetes/blob/master/cluster/common.sh#L234):

```
${0} [-M|-N|-P] -l | <release or continuous integration version> | [latest_stable|latest_release|latest_ci]
```

Furthermore, there is the script [hack/get-build.sh](https://github.com/kubernetes/kubernetes/blob/master/hack/get-build.sh), as documented in [Getting Kubernetes Builds](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/getting-builds.md).

These all need to take an arbitrary build string:

- `ci/latest`
- `ci/latest-1`
- `ci/latest-1.2`
- `release/latest`
- `release/latest-1`
- `release/latest-1.2`
- `release/stable`
- `release/stable-1`
- `release/stable-1.2`

So that we can test upgrades properly, and generally for functioning upgrades.

The logic should also be consolidated if possible.

**Comment by ihmccreery** [comment]
This is follow-on work to what I did in #15168.

**Comment by ihmccreery** [comment]
Closed #15506 via #15422.
